File: 06.js
const readline = require('readline');

const RE_INSTRUCTIONS = /(turn on|toggle|turn off) (\d+,\d+) through (\d+,\d+)/;

function interpretInstructions(lights, instructions) {
  let instructionSet = RE_INSTRUCTIONS.exec(instructions);
  if (!instructionSet) {
    process.stderr.write("uh oh, returning false");
    return false;
  }
  let directive = instructionSet[1];
  let startCoords = extractCoords(instructionSet[2].split(","));
  let endCoords = extractCoords(instructionSet[3].split(","));
  console.log(`light count was... ${countLights(lights)}`);
  var newLights;
  switch (directive) {
  case "turn on":
    newLights = turnOn(lights, startCoords, endCoords);
    break;
  case "turn off":
    newLights = turnOff(lights, startCoords, endCoords);
    break;
  case "toggle":
    newLights = toggle(lights, startCoords, endCoords);
    break;
  default:
    process.stderr.write(`uhhhh what is this? ${directive}`);
    break;
  }
  console.log(`now light count is: ${countLights(newLights)}`);
  return newLights;
}

function turnOn(l, begin, finish) {
  // console.log(`gonna turn on ${begin.y} to ${finish.y}, within every row from ${begin.x} to ${finish.x}`);
  for (let i = begin.x; i < finish.x; i++) {
    if (!l[i]) { l[i] = new Array(999); }
    for (let j = begin.y; j < finish.y; j++) {
      l[i][j] = 1;
    }
    // console.log(`done with row ${i}`);
  }
  return l;
}

function turnOff(l, begin, finish) {
  for (let i = begin.x; i < finish.x; i++) {
    if (!l[i]) { l[i] = new Array(999); }
    for (let j = begin.y; j < finish.y; j++) {
      l[i][j] = 0;
    }
  }
  return l;
}

function toggle(l, begin, finish) {
  for (let i = begin.x; i < finish.x; i++) {
    if (!l[i]) { l[i] = new Array(999); }
    for (let j = begin.y; j < finish.y; j++) {
      if (l[i][j] == 0)
        l[i][j] = 1;
      else
        l[i][j] = 0;
    }
  }
  return l;
}

function extractCoords(coords) {
  return {
    x: coords[0],
    y: coords[1]
  };
}

function countEm(sum, elem) {
  if (elem)
    return sum + 1;
  return sum;
}

function countLights(lights) {
  return lights.reduce(function(sum, lightsRow) {
    return sum + lightsRow.reduce(countEm, 0);
  }, 0);
}

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

var lines = [];

rl.on("line", function(line) {
  lines.push(line);
});

rl.on("close", function() {
  var things = [[]];
  for (var line in lines) {
    things = interpretInstructions(things, line);
  }
  console.log(countLights(things));
});

File: 06.exs
defmodule Santa do

  @re_instructions ~r/(?<directive>turn on|toggle|turn off) (?<start_coords>\d+,\d+) through (?<end_coords>\d+,\d+)/

  def read_lines(filename) do
    {:ok, contents} = File.read(filename)
    String.split(contents, "\n")
  end

  def do_it(filename\\"06head.txt") do
    filename
    |> read_lines
    |> Enum.reduce(%{}, fn
      ("", lights) -> lights
      (line, lights) ->
        interpret(line, lights)
    end)
    |> count
    |> IO.puts
  end

  def interpret(line, lights_map) do
    %{
      "directive" => directive,
      "start_coords" => start_coords,
      "end_coords" => end_coords
    } = Regex.named_captures(@re_instructions, line)
    [startX, startY] = extract_coords(start_coords)
    [endX, endY] = extract_coords(end_coords)
    IO.puts "#{count lights_map}"
    IO.write "#{directive} #{startX},#{startY} to #{endX},#{endY}... "
    Enum.reduce(startX..endX, lights_map, fn (x, lm1) ->
      Enum.reduce(startY..endY, lm1, fn (y, lm2) ->
        handle(lm2, directive, "#{x}.#{y}")
      end)
    end)
  end

  def handle(lights_map, "turn on", key) do
    previous_value = get(lights_map, key)
    Map.put(lights_map, key, previous_value + 1)
  end
  def handle(lights_map, "turn off", key) do
    previous_value = get(lights_map, key)
    new_value = if previous_value - 1 < 0, do: 0, else: previous_value - 1
    Map.put(lights_map, key, new_value)
  end
  def handle(lights_map, "toggle", key) do
    previous_value = get(lights_map, key)
    Map.put(lights_map, key, previous_value + 2)
  end

  def get(map, key), do: Map.get(map, key, 0)

  def count(lights_map) do
    Enum.reduce(lights_map, 0, fn({_key, brightness}, sum) -> sum + brightness end)
  end

  def extract_coords(coords) do
    coords
    |> String.split(",")
    |> Enum.map(&String.to_integer/1)
  end

end
